### PR TITLE
[FEATURE] TargetedSpectraExtractor: "no feature" support for all methods

### PR DESCRIPTION
**Pull request by pasqoo**
This PR implements a way to fully support usage of each of the class' methods without relying on the presence of a `FeatureMap` in input/output.

To avoid duplicating code, I decided to add a `bool` argument `compute_features` to existing methods, which decides on the computation of the `FeatureMap` in input.

I found some small details that I could improve on (like avoiding copying variables when references `&` could be used, or postponing copy of data only when really it is necessary).

In tests, I avoid the use of a global pointer and I instantiate a new `TargetedSpectraExtractor` object for each test.

### CODE REVIEW

**Comment by hroest** [comment]
seems like the test fails on windows:

```
 +  line 604:  TEST_EQUAL(selected_spectra[i].getFloatDataArrays()[1][0] >= min_score,true): got 1, expected 1
 -  line 606:  TEST_EQUAL(selected_spectra[0].getName(),"asp-L.asp-L_m2-1"): got met-L.met-L_m0-0, expected asp-L.asp-L_m2-1
 -  line 607:  TEST_REAL_SIMILAR(selected_spectra[0].getFloatDataArrays()[1][0],17.4552230834961): got 16.0294, expected 17.4552230834961 (absolute: 1.42578125000001 [1e-05], relative: 1.08894765424836 [1.00001], message: "ratio of numbers is large"
 -  line 608:  TEST_EQUAL(selected_spectra[1].getName(),"met-L.met-L_m0-0"): got asp-L.asp-L_m2-1, expected met-L.met-L_m0-0
 -  line 609:  TEST_REAL_SIMILAR(selected_spectra[1].getFloatDataArrays()[1][0],16.0294418334961): got 17.4552, expected 16.0294418334961 (absolute: 1.42578124999999 [1e-05], relative: 1.08894765424836 [1.00001], message: "ratio of numbers is large"
: failed

checking void extractSpectra( const MSExperiment& experiment, const TargetedExperiment& targeted_exp, std::vector<MSSpectrum>& extracted_spectra, FeatureMap& extracted_features, const bool compute_features = true ) const ... Warning: no default parameter description for parameters 'SavitzkyGolayFilter:frame_length,' of DefaultParameterHandler 'TargetedSpectraExtractor' given!

 +  line 634:  TEST_EQUAL(extracted_spectra.size(),extracted_features.size()): got 2, expected 2
 +  line 635:  TEST_EQUAL(extracted_spectra.size(),2): got 2, expected 2
 -  line 637:  TEST_EQUAL(extracted_spectra[0].getName(),"asp-L.asp-L_m2-1"): got met-L.met-L_m0-0, expected asp-L.asp-L_m2-1
 -  line 638:  TEST_REAL_SIMILAR(extracted_spectra[0].getFloatDataArrays()[1][0],17.4552230834961): got 16.0294, expected 17.4552230834961 (absolute: 1.42578125000001 [1e-05], relative: 1.08894765424836 [1.00001], message: "ratio of numbers is large"
 -  line 639:  TEST_EQUAL(extracted_spectra[1].getName(),"met-L.met-L_m0-0"): got asp-L.asp-L_m2-1, expected met-L.met-L_m0-0
 -  line 640:  TEST_REAL_SIMILAR(extracted_spectra[1].getFloatDataArrays()[1][0],16.0294418334961): got 17.4552, expected 16.0294418334961 (absolute: 1.42578124999999 [1e-05], relative: 1.08894765424836 [1.00001], message: "ratio of numbers is large"
: failed

checking void extractSpectra( const MSExperiment& experiment, const TargetedExperiment& targeted_exp, std::vector<MSSpectrum>& extracted_spectra ) const ... Warning: no default parameter description for parameters 'SavitzkyGolayFilter:frame_length,' of DefaultParameterHandler 'TargetedSpectraExtractor' given!

 +  line 662:  TEST_EQUAL(extracted_spectra.size(),2): got 2, expected 2
 -  line 664:  TEST_EQUAL(extracted_spectra[0].getName(),"asp-L.asp-L_m2-1"): got met-L.met-L_m0-0, expected asp-L.asp-L_m2-1
 -  line 665:  TEST_REAL_SIMILAR(extracted_spectra[0].getFloatDataArrays()[1][0],17.4552230834961): got 16.0294, expected 17.4552230834961 (absolute: 1.42578125000001 [1e-05], relative: 1.08894765424836 [1.00001], message: "ratio of numbers is large"
 -  line 666:  TEST_EQUAL(extracted_spectra[1].getName(),"met-L.met-L_m0-0"): got asp-L.asp-L_m2-1, expected met-L.met-L_m0-0
 -  line 667:  TEST_REAL_SIMILAR(extracted_spectra[1].getFloatDataArrays()[1][0],16.0294418334961): got 17.4552, expected 16.0294418334961 (absolute: 1.42578124999999 [1e-05], relative: 1.08894765424836 [1.00001], message: "ratio of numbers is large"
: failed

checking (created temporary files)...
: passed


```

**Comment by hroest** [comment]
seems like the order is not guaranteed - do you rely on uninitialized values or random values? Otherwise make sure to sort before you compare.

**Comment by pasqoo** [comment]
@hroest Thanks. I guess it happens because I'm using `unordered_map` in `selectSpectra()` and not `map`, therefore I shouldn't have relied on any sorting. Let's see what the tests say :smile:

File: src/tests/class_tests/openms/source/TargetedSpectraExtractor_test.cpp [hunk 1]
```diff
@@ -44,6 +44,20 @@
 using namespace OpenMS;
 using namespace std;
 
+vector<MSSpectrum>::const_iterator find_spectrum_by_name(const vector<MSSpectrum>& spectra, const String& name)
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/TargetedSpectraExtractor_test.cpp]: chagne to findSpectrumByName

Comment (pasqoo) [reply src/tests/class_tests/openms/source/TargetedSpectraExtractor_test.cpp]: Is it the preferred way of naming non-class functions? @timosachsenberg

Comment (timosachsenberg) [reply src/tests/class_tests/openms/source/TargetedSpectraExtractor_test.cpp]: I think so. At least we also use that for static functions...

**Comment by pasqoo** [comment]
```
Build project
   Each symbol represents 1024 bytes of output.
   '!' represents an error and '*' a warning.
    ..***********************.*.****************.*....  Size: 49K
    ...*********..*.****.*****.*.**.******************  Size: 99K
    ******...**.*********.*******.*****..*.***********  Size: 149K
    **************************************************  Size: 199K
    *****.****..**.***.****************.**********.***  Size: 249K
    *..***.**.

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```
@hroest @timosachsenberg previous tests look fine, though

**Comment by timosachsenberg** [comment]
let's restart... probably just a hickup

**Pull request merged**